bgpd/bgp_vty.c

```diff
@@ -7877,70 +7877,52 @@ ALIAS_HIDDEN(neighbor_advertise_map, neighbor_advertise_map_hidden_cmd,
 	     "Name of the exist or non exist map\n")
 
 /* Set route-map to the peer. */
-static int peer_route_map_set_vty(struct vty *vty, const char *ip_str,
-				  afi_t afi, safi_t safi, const char *name_str,
-				  const char *direct_str)
-{
-	int ret;
-	struct peer *peer;
-	int direct = RMAP_IN;
-	struct route_map *route_map;
-
-	peer = peer_and_group_lookup_vty(vty, ip_str);
-	if (!peer)
-		return CMD_WARNING_CONFIG_FAILED;
-
-	/* Check filter direction. */
-	if (strncmp(direct_str, "in", 2) == 0)
-		direct = RMAP_IN;
-	else if (strncmp(direct_str, "o", 1) == 0)
-		direct = RMAP_OUT;
-
-	route_map = route_map_lookup_warn_noexist(vty, name_str);
-	ret = peer_route_map_set(peer, afi, safi, direct, name_str, route_map);
-
-	return bgp_vty_return(vty, ret);
-}
-
-static int peer_route_map_unset_vty(struct vty *vty, const char *ip_str,
-				    afi_t afi, safi_t safi,
-				    const char *direct_str)
+DEFPY_YANG(
+	neighbor_route_map, neighbor_route_map_cmd,
+	"[no$no] neighbor <A.B.C.D|X:X::X:X|WORD>$neighbor_str route-map WORD$rmap_str  <in|out>$direction",
+	NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+	"Apply route map to neighbor\n"
+	"Name of route map\n"
+	"Apply map to incoming routes\n"
+	"Apply map to outbound routes\n")
 {
-	int ret;
-	struct peer *peer;
-	int direct = RMAP_IN;
+	char base_xpath[XPATH_MAXLEN];
+	char af_xpath[XPATH_MAXLEN];
+	char rmap_xpath[80];
+	afi_t afi = bgp_node_afi(vty);
+	safi_t safi = bgp_node_safi(vty);
 
-	peer = peer_and_group_lookup_vty(vty, ip_str);
-	if (!peer)
+	snprintf(af_xpath, sizeof(af_xpath), FRR_BGP_AF_XPATH,
+		 yang_afi_safi_value2identity(afi, safi));
+	if (peer_and_group_lookup_nb(vty, neighbor_str, base_xpath,
+				     sizeof(base_xpath), af_xpath)
+	    < 0)
 		return CMD_WARNING_CONFIG_FAILED;
 
-	/* Check filter direction. */
-	if (strncmp(direct_str, "in", 2) == 0)
-		direct = RMAP_IN;
-	else if (strncmp(direct_str, "o", 1) == 0)
-		direct = RMAP_OUT;
-
-	ret = peer_route_map_unset(peer, afi, safi, direct);
+	if (strmatch(direction, "in"))
+		snprintf(rmap_xpath, sizeof(rmap_xpath),
+			 "./%s/filter-config/rmap-import",
+			 bgp_afi_safi_get_container_str(afi, safi));
+	else if (strmatch(direction, "out"))
+		snprintf(rmap_xpath, sizeof(rmap_xpath),
+			 "./%s/filter-config/rmap-export",
+			 bgp_afi_safi_get_container_str(afi, safi));
 
-	return bgp_vty_return(vty, ret);
-}
+	if (!no) {
+		route_map_lookup_warn_noexist(vty, rmap_str);
```

Comment:
May i know, why you are accessing the backend datastructs in CLI context ?
One reason could be rmap PR is still open. Are you going to modify this call after rmap PR merged ?

Reply:
Ack. I have modified to lookup route-map data tree to check if dnode exists.